test/tests/parser_tests.py

```diff
@@ -0,0 +1,305 @@
+"""Tests for the various Pavilion parsers."""
+
+import lark
+from pavilion import plugins
+from pavilion import unittest
+from pavilion.test_config import parsers
+from pavilion.test_config import variables
+from pavilion import system_variables
+
+
+class ParserTests(unittest.PavTestCase):
+
+    def setUp(self) -> None:
+        plugins.initialize_plugins(self.pav_cfg)
+
+        self.var_man = variables.VariableSetManager()
+        self.var_man.add_var_set('var', {
+            'int1': "1",
+            'int2': "2",
+            'float1': '1.1',
+            'str1': 'hello',
+            'ints': ['0', '1', '2', '3', '4', '5'],
+            'floats': ['0.1', '2.3'],
+            'more_ints': ['0', '1'],
+            'struct': {
+                'cpus': '200',
+                'flops': '2.1',
+                'name': 'earth_chicken',
+            },
+            'structs': [
+                {'type': 'cat', 'bites': '3', 'evil_rating': '5.2'},
+                {'type': 'dog', 'bites': '0', 'evil_rating': '0.2'},
+                {'type': 'fish', 'bites': '1', 'evil_rating': '9.7'},
+            ]
+        })
+        self.var_man.add_var_set('sys', system_variables.get_vars(defer=True))
+
+    def tearDown(self) -> None:
+        plugins._reset_plugins()
+
+    def test_visitors(self):
+        """Test the visitors used to collect all of the used variables."""
+
+        expr_parser = parsers.get_expr_parser()
+
+        expr = 'int1.3.foo + var.int2.*.bleh * 11 * - sum([int1, int1])'
+        tree = expr_parser.parse(expr)
+        visitor = parsers.common.VarRefVisitor()
+        used_vars = visitor.visit(tree)
+
+        self.assertEqual(used_vars,
+                         ['int1.3.foo', 'var.int2.*.bleh', 'int1'])
+
+        string_parser = parsers.get_string_parser()
+        tree = string_parser.parse("hello {{var.foo.1.baz:1s}} world "
+                                   "{{sys.bar}}")
+        visitor = parsers.strings.StringVarRefVisitor()
+        var_list = visitor.visit(tree)
+        self.assertEqual(['var.foo.1.baz', 'sys.bar'], var_list)
+
+    GOOD_EXPRESSIONS = {
```

Comment:
I don't see any normal division (non-floor), multiplication by zero, or division by 0.  Presumably that last one isn't a 'GOOD_EXPRESSIONS', but I also don't see it in the bad group below.

Reply:
That's a damn good catch.